File: experiments/supervisor/supervisor.py
"""Minimal ACP supervisor — spawns claude-agent-acp, bridges stdio ↔ WebSocket.

One subprocess per supervisor instance. One WebSocket client at a time (this
is a spike, not production). Client sends JSON-RPC frames as WS text; we
write them to subprocess stdin. Subprocess stdout lines are streamed back
to the client as WS text. stderr goes to our own stderr.

Run:
    python supervisor.py --port 9000 --acp /path/to/claude-agent-acp

WebSocket endpoint: ws://host:9000/acp
"""
from __future__ import annotations

import argparse
import asyncio
import logging
import os
import sys
from contextlib import asynccontextmanager
from typing import Optional

from fastapi import FastAPI, WebSocket, WebSocketDisconnect
from fastapi.responses import JSONResponse
import uvicorn

log = logging.getLogger("supervisor")

# Global state (spike — one subprocess per supervisor)
_state: dict = {
    "proc": None,
    "ws": None,
    "reader_task": None,
    "stderr_task": None,
    "acp_bin": None,
}


async def _start_acp() -> asyncio.subprocess.Process:
    bin_path = _state["acp_bin"]
    if not bin_path:
        raise RuntimeError("acp binary path not configured")
    env = os.environ.copy()
    proc = await asyncio.create_subprocess_exec(
        bin_path,
        stdin=asyncio.subprocess.PIPE,
        stdout=asyncio.subprocess.PIPE,
        stderr=asyncio.subprocess.PIPE,
        env=env,
        cwd="/tmp",
    )
    log.info("spawned claude-agent-acp pid=%d", proc.pid)
    return proc


async def _forward_stdout_to_ws(proc: asyncio.subprocess.Process) -> None:
    """Read ndjson lines from subprocess stdout, forward to WS."""
    try:
        while True:
            line = await proc.stdout.readline()
            if not line:
                break
            ws = _state["ws"]
            if ws is None:
                log.warning("no ws client; dropping line: %s", line[:80])
                continue
            try:
                await ws.send_text(line.decode(errors="replace"))
            except Exception as e:
                log.warning("ws send failed: %s", e)
                break
    except Exception as e:
        log.exception("stdout forward error: %s", e)


async def _forward_stderr(proc: asyncio.subprocess.Process) -> None:
    try:
        while True:
            line = await proc.stderr.readline()
            if not line:
                break
            sys.stderr.write(f"[acp-stderr] {line.decode(errors='replace')}")
            sys.stderr.flush()
    except Exception as e:
        log.exception("stderr forward error: %s", e)


@asynccontextmanager
async def lifespan(app: FastAPI):
    # Start acp process at supervisor startup, keep it alive for the whole
    # lifetime of the supervisor. One supervisor = one ACP connection.
    proc = await _start_acp()
    _state["proc"] = proc
    _state["reader_task"] = asyncio.create_task(
        _forward_stdout_to_ws(proc), name="acp-stdout-forward"
    )
    _state["stderr_task"] = asyncio.create_task(
        _forward_stderr(proc), name="acp-stderr-forward"
    )
    yield
    # Shutdown
    for t in (_state["reader_task"], _state["stderr_task"]):
        if t and not t.done():
            t.cancel()
            try:
                await t
            except (asyncio.CancelledError, Exception):
                pass
    if proc.returncode is None:
        try:
            proc.stdin.close()
            await asyncio.wait_for(proc.wait(), timeout=5)
        except Exception:
            proc.kill()
            await proc.wait()


app = FastAPI(lifespan=lifespan)


@app.get("/health")
async def health():
    proc = _state["proc"]
    return {
        "status": "ok",
        "acp_pid": proc.pid if proc else None,
        "acp_alive": (proc is not None and proc.returncode is None),
        "ws_connected": _state["ws"] is not None,
    }


@app.websocket("/acp")
async def acp_ws(ws: WebSocket):
    await ws.accept()
    if _state["ws"] is not None:
        await ws.close(code=4000, reason="another client already connected")
        return
    _state["ws"] = ws
    proc = _state["proc"]
    if not proc or proc.returncode is not None:
        await ws.close(code=4001, reason="acp subprocess not alive")
        _state["ws"] = None
        return
    log.info("ws client connected")
    try:
        while True:
            msg = await ws.receive_text()
            line = (msg.rstrip("\n") + "\n").encode()
            proc.stdin.write(line)
            await proc.stdin.drain()
    except WebSocketDisconnect:
        log.info("ws client disconnected")
    except Exception as e:
        log.exception("ws loop error: %s", e)
    finally:
        if _state["ws"] is ws:
            _state["ws"] = None


def main():
    parser = argparse.ArgumentParser()
    parser.add_argument("--host", default="0.0.0.0")
    parser.add_argument("--port", type=int, default=9000)
    parser.add_argument("--acp", required=True, help="path to claude-agent-acp binary or node entry")
    args = parser.parse_args()

    _state["acp_bin"] = args.acp

    logging.basicConfig(
        level=logging.INFO,
        format="%(asctime)s [%(levelname)s] %(message)s",
        stream=sys.stderr,
    )
    log.info("supervisor starting on %s:%d acp=%s", args.host, args.port, args.acp)
    uvicorn.run(app, host=args.host, port=args.port, log_level="warning")


if __name__ == "__main__":
    main()

File: experiments/supervisor/supervisor.js
#!/usr/bin/env node
/**
 * Minimal ACP supervisor — spawns claude-agent-acp, bridges stdio ↔ WebSocket.
 *
 * Usage: node supervisor.js --port 9100 --acp /path/to/claude-agent-acp
 *
 * HTTP endpoints:
 *   GET  /health      → {status, acp_pid, ws_connected}
 *   WS   /acp         → text frames are JSON-RPC; bidirectional bridge
 *
 * One ws client at a time. For spike use only.
 */
const http = require("node:http");
const { spawn } = require("node:child_process");
const { WebSocketServer } = require("ws");

function parseArgs(argv) {
    const out = { port: 9100, acp: null, host: "0.0.0.0" };
    for (let i = 2; i < argv.length; i++) {
        const a = argv[i];
        if (a === "--port") out.port = parseInt(argv[++i], 10);
        else if (a === "--host") out.host = argv[++i];
        else if (a === "--acp") out.acp = argv[++i];
    }
    if (!out.acp) {
        console.error("--acp required");
        process.exit(1);
    }
    return out;
}

function log(...args) {
    const t = new Date().toISOString();
    process.stderr.write(`${t} [supervisor] ${args.join(" ")}\n`);
}

const args = parseArgs(process.argv);

// Spawn claude-agent-acp once at startup; keep it alive for the supervisor lifetime.
const acp = spawn(args.acp, [], {
    stdio: ["pipe", "pipe", "pipe"],
    env: process.env,
    cwd: "/tmp",
});
log("spawned acp pid=" + acp.pid);

acp.stderr.on("data", (chunk) => {
    process.stderr.write("[acp-stderr] " + chunk.toString());
});

acp.on("exit", (code, signal) => {
    log(`acp exited code=${code} signal=${signal}`);
    process.exit(code || 1);
});

// One ws client at a time.
let currentWs = null;

// Forward acp stdout lines to the current ws (line-delimited JSON-RPC).
let stdoutBuf = "";
acp.stdout.on("data", (chunk) => {
    stdoutBuf += chunk.toString("utf8");
    let idx;
    while ((idx = stdoutBuf.indexOf("\n")) !== -1) {
        const line = stdoutBuf.slice(0, idx);
        stdoutBuf = stdoutBuf.slice(idx + 1);
        if (!line) continue;
        if (currentWs && currentWs.readyState === currentWs.OPEN) {
            currentWs.send(line);
        } else {
            // No client — drop. Could buffer but for the spike we don't.
        }
    }
});

const server = http.createServer((req, res) => {
    if (req.url === "/health") {
        const body = JSON.stringify({
            status: "ok",
            acp_pid: acp.pid,
            acp_alive: acp.exitCode === null,
            ws_connected: currentWs !== null,
        });
        res.writeHead(200, { "content-type": "application/json" });
        res.end(body);
        return;
    }
    res.writeHead(404, { "content-type": "text/plain" });
    res.end("not found\n");
});

const wss = new WebSocketServer({ server, path: "/acp" });
wss.on("connection", (ws) => {
    if (currentWs !== null) {
        ws.close(4000, "another client already connected");
        return;
    }
    currentWs = ws;
    log("ws client connected");

    ws.on("message", (data, isBinary) => {
        const text = isBinary ? data.toString("utf8") : data.toString();
        // Ensure single-line frame terminated with \n for the acp stdin reader.
        const line = text.replace(/\n+$/, "") + "\n";
        try {
            acp.stdin.write(line);
        } catch (e) {
            log("acp stdin write failed: " + e.message);
            ws.close(4002, "acp stdin error");
        }
    });

    ws.on("close", () => {
        if (currentWs === ws) currentWs = null;
        log("ws client disconnected");
    });

    ws.on("error", (e) => {
        log("ws error: " + e.message);
    });
});

server.listen(args.port, args.host, () => {
    log(`listening on ${args.host}:${args.port}`);
});

function shutdown() {
    log("shutting down");
    try { acp.stdin.end(); } catch {}
    try { server.close(); } catch {}
    setTimeout(() => process.exit(0), 500);
}
process.on("SIGTERM", shutdown);
process.on("SIGINT", shutdown);
